jetty-util/src/main/java/org/eclipse/jetty/util/Index.java

```diff
@@ -0,0 +1,340 @@
+         * @param v The value of the entry
+         * @return True if the index had capacity to add the field.
+         */
+        boolean put(String s, V v);
+
+        /**
+         * Put a value as both a key and a value.
+         *
+         * @param v The value and key
+         * @return True if the Trie had capacity to add the field.
+         */
+        boolean put(V v);
+
+        /**
+         * Remove an entry from the index.
+         *
+         * @param s The key for the entry
+         * @return The removed value of the entry
+         */
+        V remove(String s);
+
+        /**
+         * Remove all entries from the index.
+         */
+        void clear();
+
+        /**
+         * Builder of {@link Index.Mutable} instances. Such builder cannot be
+         * directly created, it is instead returned by calling {@link Index.Builder#mutable()}.
+         * @param <V> the entry type
+         */
+        class Builder<V> extends Index.Builder<V>
+        {
+            private int maxCapacity = -1;
+
+            Builder(Map<String, V> contents, boolean caseSensitive)
+            {
+                super(contents, caseSensitive);
+            }
+
+            /**
+             * Configure a maximum capacity for the mutable index.
+             * A negative value means there is no capacity limit and
+             * the index can grow without limits.
+             * The default value is -1.
+             * @param capacity the maximum capacity of the index.
+             * @return this
+             */
+            public Builder<V> maxCapacity(int capacity)
+            {
+                this.maxCapacity = capacity;
+                return this;
+            }
+
+            /**
+             * Build a {@link Mutable} instance.
+             * @return a {@link Mutable} instance.
+             */
+            public Mutable<V> build()
+            {
+                Mutable<V> result;
+                if (maxCapacity > 0)
+                    result = new ArrayTernaryTrie<>(!caseSensitive, maxCapacity);
+                else if (maxCapacity < 0)
+                    result = new ArrayTernaryTrie.Growing<>(!caseSensitive, 512, 512);
+                else
+                    return EmptyTrie.instance(caseSensitive);
+
+                for (Map.Entry<String, V> entry : contents.entrySet())
+                {
+                    if (!result.put(entry.getKey(), entry.getValue()))
+                        throw new AssertionError("Trie capacity exceeded at " + entry.getKey());
+                }
+                return result;
+            }
+        }
+    }
+
+    /**
+     * Builder of {@link Index} instances.
+     * @param <V> the entry type
+     */
+    class Builder<V>
+    {
+        Map<String, V> contents;
+        boolean caseSensitive;
+
+        /**
+         * Create a new index builder instance.
+         */
+        public Builder()
+        {
+            this(new LinkedHashMap<>(), false);
```

Comment:
We need to create this lazily only when a `with` method is called....
either that or we need a cheaper way for HttpParser to create a cache index on the fly without creating a lot of garbage.
Currently that code is:
```java
                            ? new Index.Builder<HttpField>()
                            .caseSensitive(false)
                            .mutable()
                            .maxCapacity(getHeaderCacheSize())
                            .build()
```
So even with a lazy contents it is creating two `Builders` with an `int` field in one and a `boolean` plus int in the other...
Perhaps it would be better to support a short-cut like:

```java
                            ? new Index.MutableBuilder<HttpField>()
                            .caseSensitive(false)
                            .maxCapacity(getHeaderCacheSize())
                            .build()
```

Reply:
I could make the `Index.MutableBuilder` constructor public, but if I make the `HashMap`s lazy I don't think cutting a single object that holds an `int` will save anything measurable.

Reply:
I think it is still worth making the constructor public... even a small cost is significant in the HttpParser